File: hard/Lc212.py
"""
2022年4月8日

"""
class Node:
    def __init__(self):
        self.chs = {}
        self.word = ""

class Trie:
    def __init__(self):
        self.root = Node()

    def add(self, word):
        cur = self.root
        for w in word:
            if w not in cur.chs:
                cur.chs[w] = Node()
            cur = cur.chs[w]
        cur.word = word

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        m, n = len(board), len(board[0])
        ans = []
        trie = Trie()
        for word in words:
            trie.add(word)

        def dfs(i, j, board, node):
            nonlocal m, n
            if i < 0 or i >= m or j < 0 or j >= n:
                return
            ch = board[i][j]
            if ch == "#":
                return
            if ch not in node.chs:
                return
            node = node.chs[ch]
            if node.word != "" and node.word not in ans:
                ans.append(node.word)
            board[i][j] = "#"
            dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
            for dir in dirs:
                dfs(i + dir[0], j + dir[1], board, node)
            board[i][j] = ch

        for i in range(m):
            for j in range(n):
                dfs(i, j, board, trie.root)
        return ans

File: hard/Lc212/Lc212.go
package Lc212

import . "github.com/xiaozhuoops/leetcode/template/trie"

//dfs+trie
//i,j node,k dfs()
//special case: "ab" "abc"
type pos = [2]int

func findWords(board [][]byte, words []string) []string {
	// build trie
	trie := Trie{}
	for _, word := range words {
		trie.Insert(word)
	}
	// dfs
	ans := make(map[string]struct{})
	m, n := len(board), len(board[0])
	dirs := []pos{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	vis := make(map[pos]bool)

	var dfs func(x int, y int, node *Trie)
	dfs = func(x int, y int, par *Trie) {
		if 0 > x || 0 > y || x >= m || y >= n || vis[pos{x, y}] {
			return
		}
		node := par.Chs[board[x][y]-'a']
		if node == nil {
			return
		} else if node.Word != "" {
			ans[node.Word] = struct{}{}
			//return
		}
		vis[pos{x, y}] = true
		for _, dir := range dirs {
			dfs(x+dir[0], y+dir[1], node)
		}
		vis[pos{x, y}] = false
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			dfs(i, j, &trie)
		}
	}
	var ansList []string
	for k := range ans {
		ansList = append(ansList, k)
	}
	return ansList
}
